src/librustc_trait_selection/traits/select/mod.rs

```diff
@@ -2124,7 +2124,7 @@ impl<'tcx> TraitObligationExt<'tcx> for TraitObligation<'tcx> {
         // by using -Z verbose or just a CLI argument.
         let derived_cause = DerivedObligationCause {
             parent_trait_ref: obligation.predicate.to_poly_trait_ref(),
-            parent_code: Rc::new(obligation.cause.code.clone()),
+            parent_code: Rc::new(obligation.cause.code().clone()),
```

Comment:
```suggestion
            parent_code: obligation.cause.code(),
```
And also change `parent_code` to `Option<Rc<ObligationCauseCode>>`

Reply:
in general, all occurences of `parent_code` currently reallocate